app/src/docs/content/streams/endToEndEncryption.mdx

```diff
@@ -1,19 +1,30 @@
 
-Currently, the symmetric key must be shared between the publisher and the subscribers in a secure way outside of the Streamr Network. The publisher can pass group keys per stream as parameters and the subscriber can pass group keys per stream and per publisher as parameters. Later it will be possible to use a secure key-exchange protocol directly in the Streamr Network.
+Confidentiality of events published on a stream can be guaranteed with end-to-end encryption. The publisher generates a AES-256 symmetric encryption key and encrypts the messages before publishing them to the network. As the publisher fully controls who can access their data, they are also responsible for communicating the key to subscribers - usually via the key exchange mechanism described below.
 
-In order to prevent new subscribers to eavesdrop and then decrypt past messages published before the start of their subscription, the publisher can decide at any time to update the symmetric group key by sending a new key concatenated with an event and encrypted with the current key:
-```
-// normal case when creating a ciphertext of an event 'm1' with a key 'k1'
-c1 = AES_encrypt(m1, k1)
-// when updating the key to 'k2', current subscribers will recover 'k2'
-c2 = AES_encrypt(m2 + k2, k1)
-// the next messages can be encrypted with 'k2', new subscribers obtain 'k2' but not 'k1' so they cannot decrypt m1 and m2
-c3 = AES_encrypt(m3, k2)
-```
+## Key exchange
 
-Both the Javascript and the Java SDKs support this key update mechanism. It can be triggered by calling the `publish` method with a new symmetric group key as parameter.
+The subscribers need this symmetric group key in order to decrypt the data. They automatically obtain the key by performing a key exchange with the publisher. The key exchange happens using asymmetric encryption: 
+
+- Both the publisher and subscriber generate a temporary RSA key pair to be used for the key exchange
+- The subscriber sends a key request to the publisher, containing the subscriber's public key, signed with the subscriber's Ethereum key
+- The publisher checks from the on-chain access control registry whether that subscriber should be able to access the stream, and if it does, the publisher responds with the AES symmetric key, encrypted with the publisher's RSA key for the subscriber's RSA key, and signs with the publisher's Ethereum key.
+
+## Publisher liveness
+
+To perform key exchange with subscribers, the publisher must be online and present in the network. As the Streamr Network deals with real-time messages, publishers are often constantly online. Especially with old stored data it may happen, however, that the publisher has disappeared since publishing the data, making those messages inaccessible to subscribers who didn't receive the key at the time. This is a consequence of publishers being fully and solely in control of who can access their data on the network.
```

Comment:
```suggestion
To perform key exchange with subscribers, the publisher must be online and present in the Network. As the Streamr Network deals with real-time messages, publishers are often constantly online. However, it may happen that the publisher has disappeared since publishing the data, making those messages inaccessible to subscribers who have yet to receive the key. This is a consequence of the data publishers being fully in control of who can access their data on the Network.
```